### PR TITLE
HPCC-20426 cryptohelper into jlib dll, use openssl routines where available, refactor digisign

### PR DESCRIPTION
**Pull request by jakesmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - based off hpcc-20383. Please take a look at last 2 commits.

**Comment by jakesmith** [comment]
@ghalliday - I managed to build under windows and have resolved the build issue.

This PR also contains some refactoring of the digisign, to avoid some of threading issues and be more efficient + common up the setting up of ssl callbacks for locking etc.
So this PR also resolves - HPCC-20345

Please review and double check windows compilation.

**Review by ghalliday — COMMENTED** [review]
@jakesmith a few comments.  Mostly reinstating tests that have been removed, and other minor comments.

File: testing/unittests/cryptotests.cpp [hunk 1]
```diff
@@ -85,11 +85,11 @@ class CryptoUnitTest : public CppUnit::TestFixture
 public:
     CPPUNIT_TEST_SUITE(CryptoUnitTest);
         CPPUNIT_TEST(digiSignTests);
-        CPPUNIT_TEST(pkeEncryptDecryptTest);
-        CPPUNIT_TEST(pkeParallelTest);
-        CPPUNIT_TEST(aesEncryptDecryptTests);
-        CPPUNIT_TEST(aesWithRsaEncryptedKey);
-        CPPUNIT_TEST(aesParallelTest);
+//        CPPUNIT_TEST(pkeEncryptDecryptTest);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply testing/unittests/cryptotests.cpp]: whoops yes

File: testing/unittests/cryptotests.cpp [hunk 2]
```diff
@@ -201,90 +208,90 @@ class CryptoUnitTest : public CppUnit::TestFixture
             printf("\nExecuting digiSign() unit tests\n");
 
             //Create instance of digital signature manager
-            StringBuffer _pubKeyBuff(pubKey);
-            StringBuffer _privKeyBuff(privKey);
-            Owned<IDigitalSignatureManager> dsm(createDigitalSignatureManagerInstanceFromKeys(_pubKeyBuff, _privKeyBuff, nullptr));
-
+            Owned<IDigitalSignatureManager> dsm(createDigitalSignatureManagerInstanceFromKeys(pubKey, privKey, nullptr));
 
+/*
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply testing/unittests/cryptotests.cpp]: will do

File: system/jlib/jencrypt.cpp [hunk 3]
```diff
@@ -1798,8 +1804,29 @@ MemoryBuffer &aesDecrypt(const void *key, size_t keylen, const void *input, size
...
         throw MakeStringException(-1,"AES Decryption error: %d, %s", len, getAesErrorText(len));
+    return output;
+}
 
+} // end of namespace jlib
+
+MemoryBuffer &aesEncrypt(const void *key, size_t keylen, const void *input, size_t inlen, MemoryBuffer &output)
+{
+#if defined(_USE_OPENSSL) && !defined(_WIN32)
+	cryptohelper::aesEncrypt(output, inlen, input, keylen, (const char *)key);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/jlib/jencrypt.cpp]: it's a tab, not sure how that got there, will fix.

File: system/security/cryptohelper/digisign.cpp [hunk 4]
```diff
@@ -277,79 +194,101 @@ static void createDigitalSignatureManagerInstance(IDigitalSignatureManager * * p
...
-            {
-                e->Release();
-            }
-            if (publicKeyBuff.isEmpty())
-                throw MakeStringException(-1, "digiSign:Cannot load public key file");
+            privKey.setown(loadPrivateKeyFromFile(privKeyFileName, passPhrase));
         }
-
-        if (!isEmptyString(_privKey))
+        catch (IException * e)
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/security/cryptohelper/digisign.cpp]: I don't really like the IDigitalSignatureManager conflating the two. I introduced separate digiSign and digiVerify standalone methods which is what HPCC-20340 will use (digiSign in esp and digiVerify in dafilesrv).

I will change it report the error but allow it to continue.

File: system/security/cryptohelper/digisign.hpp [hunk 5]
```diff
@@ -27,31 +27,41 @@
...
 {
 public:
-    virtual bool isDigiSignerConfigured() = 0;
-    virtual bool isDigiVerifierConfigured() = 0;
-    virtual bool digiSign(const char * text, StringBuffer & b64Signature) = 0;//signs, using private key
-    virtual bool digiVerify(const char * text, StringBuffer & b64Signature) = 0;//verifies, using public key
+    virtual bool isDigiSignerConfigured() const = 0;
+    virtual bool isDigiVerifierConfigured() const = 0;
+    virtual bool digiSign(StringBuffer & b64Signature, const char * text) const = 0;//signs, using private key
+    virtual bool digiVerify(StringBuffer & b64Signature, const char * text) const = 0;//verifies, using public key
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/security/cryptohelper/digisign.hpp]: yeah I agree, I will change.

File: system/security/cryptohelper/cryptocommon.cpp [hunk 6]
```diff
@@ -17,21 +17,51 @@
...
+}
+
+static void initSSLLibrary()
+{
+    SSL_load_error_strings();
+    SSLeay_add_ssl_algorithms();
+    numCryptoLocks = CRYPTO_num_locks();
+    for (unsigned i=0; i<numCryptoLocks; i++)
+        cryptoLocks.push_back(std::unique_ptr<CriticalSection>(new CriticalSection));
+    CRYPTO_set_locking_callback(locking_function);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/security/cryptohelper/cryptocommon.cpp]: have changed.

File: system/security/cryptohelper/digisign.hpp [hunk 7]
```diff
@@ -27,31 +27,41 @@
...
+#include "pke.hpp"
+
 namespace cryptohelper
 {
 
+//Create base 64 encoded digital signature of given text string
+bool digiSign(StringBuffer &b64Signature, const char *text, const CLoadedKey &signingKey);
+
+//Verify the given text was used to create the given digital signature
+bool digiVerify(StringBuffer &b64Signature, const char *text, const CLoadedKey &verifyingKey);
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply system/security/cryptohelper/digisign.hpp]: agreed, will change

**Comment by jakesmith** [comment]
@ghalliday - please see review changes.

**Comment by ghalliday** [comment]
@jakesmith I think you haven't reversed the parameters on the two calls to digiVerify in daldap.cpp

**Comment by ghalliday** [comment]
Other than the comment above it looks ready to merge.  Please can you squash once you have addressed the verify calls.

Note the changes do add a noticeable improvement on the digiSign/digiVerify timings.
On my machine (only 1 run) Signing 1941 -> 1474  Verify 57->41, async signing improved even more.

**Comment by jakesmith** [comment]
@ghalliday - please merge after smoketest passes.

**Comment by AttilaVamos** [comment]
@jakesmith The last commit: 895f504 won't be tested, because the PR is merged/closed.

**Comment by jakesmith** [comment]
@AttilaVamos - that's fine - it was a WIN32 addition only, so smoketest wouldn't see any difference anyway.

**Comment by AttilaVamos** [comment]
@jakesmith OK, thanks

**Pull request merged**